test/python/transpiler/test_remove_final_measurements.py

```diff
@@ -0,0 +1,221 @@
+        qc.measure(1, c0[1])
+
+        dag = circuit_to_dag(qc)
+        remove_final_meas = RemoveFinalMeasurements()
+        remove_final_meas.run(dag)
+
+        self.assertFalse(dag.cregs)
+        self.assertFalse(dag.clbits)
+
+    def test_final_measures_share_dest(self):
+        """Multiple final measurements use the same clbit."""
+
+        def expected_dag():
+            qc = QuantumCircuit(QuantumRegister(2, "q0"))
+            return circuit_to_dag(qc)
+
+        rq = QuantumRegister(2, "q0")
+        rc = ClassicalRegister(1, "c0")
+        qc = QuantumCircuit(rq, rc)
+
+        qc.measure(0, 0)
+        qc.measure(1, 0)
+
+        dag = circuit_to_dag(qc)
+        RemoveFinalMeasurements().run(dag)
+
+        self.assertEqual(dag, expected_dag())
+
+    def test_remove_chained_final_measurements(self):
+        """Remove successive final measurements."""
+
+        def expected_dag():
+            q0 = QuantumRegister(1, "q0")
+            q1 = QuantumRegister(1, "q1")
+            c0 = ClassicalRegister(1, "c0")
+            qc = QuantumCircuit(q0, c0, q1)
+
+            qc.measure(q0, c0)
+            qc.measure(q0, c0)
+            qc.barrier()
+            qc.h(q1)
+
+            return circuit_to_dag(qc)
+
+        q0 = QuantumRegister(1, "q0")
+        q1 = QuantumRegister(1, "q1")
+        c0 = ClassicalRegister(1, "c0")
+        c1 = ClassicalRegister(1, "c1")
+        qc = QuantumCircuit(q0, c0, q1, c1)
+
+        qc.measure(q0, c0)
+        qc.measure(q0, c0)
+        qc.barrier()
+        qc.h(q1)
+        qc.measure(q1, c1)
+        qc.measure(q0, c1)
+
+        dag = circuit_to_dag(qc)
+        RemoveFinalMeasurements().run(dag)
+
+        self.assertEqual(dag, expected_dag())
+
+    def test_remove_clbits_without_register(self):
+        """clbits of final measurements not in a register are removed."""
+        qc = QuantumCircuit(1)
+
+        # Add clbit without adding register
+        clbit = Clbit(ClassicalRegister(1), 0)
+        qc.add_bits([clbit])
+
+        self.assertFalse(qc.cregs)
+
+        # Measure to regless clbit
+        qc.measure(0, 0)
+
+        dag = circuit_to_dag(qc)
+        RemoveFinalMeasurements().run(dag)
+
+        self.assertFalse(dag.cregs)
+        self.assertFalse(dag.clbits)
+
+    def test_final_barriers_and_measures_complex(self):
+        """Test complex final barrier and measure removal."""
+
+        def expected_dag():
+            q0 = QuantumRegister(5, "q0")
+            c1 = ClassicalRegister(1, "c1")
+            qc = QuantumCircuit(q0, c1)
+            qc.h(q0[0])
+            return circuit_to_dag(qc)
+
+        q0 = QuantumRegister(5, "q0")
+        c0 = ClassicalRegister(1, "c0")
+        c1 = ClassicalRegister(1, "c1")
+        qc = QuantumCircuit(q0, c0, c1)
+
+        qc.measure(q0[1], c0)
+        qc.h(q0[0])
+        qc.measure(q0[0], c0[0])
+        qc.barrier()
+        qc.barrier(q0[2], q0[3])
+        qc.measure_all()
+        qc.barrier(q0[4])
```

Comment:
Hmm, I'm not actually sure even I know what you're referring to :). The expected circuit here only has a single `H` operation. Even though the circuit ends with a barrier for `q0[4]`, a barrier with no non-final children is (at least according to the original doc string of the `RemoveFinalMeasurements` pass) a final operation. So the pass name itself is a bit of a misnomer. A better name would be `RemoveFinalMeasurementsAndBarriers`. Unless of course the original doc string wasn't correct. Let me know if I should make changes 😅.

Reply:
Not at all, I just mis-read the test case and thought it was testing something else :).

The test circuit here looks like:
```
        ┌───┐┌─┐ ░     ░ ┌─┐               
  q0_0: ┤ H ├┤M├─░─────░─┤M├───────────────
        └┬─┬┘└╥┘ ░     ░ └╥┘┌─┐            
  q0_1: ─┤M├──╫──░─────░──╫─┤M├────────────
         └╥┘  ║  ░  ░  ░  ║ └╥┘┌─┐         
  q0_2: ──╫───╫──░──░──░──╫──╫─┤M├─────────
          ║   ║  ░  ░  ░  ║  ║ └╥┘┌─┐      
  q0_3: ──╫───╫──░──░──░──╫──╫──╫─┤M├──────
          ║   ║  ░  ░  ░  ║  ║  ║ └╥┘┌─┐ ░ 
  q0_4: ──╫───╫──░─────░──╫──╫──╫──╫─┤M├─░─
          ║   ║  ░     ░  ║  ║  ║  ║ └╥┘ ░ 
  c0: 1/══╩═══╩═══════════╬══╬══╬══╬══╬════
          0   0           ║  ║  ║  ║  ║    
                          ║  ║  ║  ║  ║    
  c1: 1/══════════════════╬══╬══╬══╬══╬════
                          ║  ║  ║  ║  ║    
meas: 5/══════════════════╩══╩══╩══╩══╩════
                          0  1  2  3  4    
​
```
I mistook the first `measure` to also be on `q0[0]` before the `h` (so that, if one followed only along `Clbit` edges, it would look like a final operation,  ), but this would also lead to a different output, so feel free to disregard.

Agree the pass name is a bit of a misnomer (though this has been the case since it was introduced in #2993 ). For now, I think the best path forward here is to make the behavior and documentation consistent, leaving the nameing as is.